system-addon/content-src/components/Sections/Sections.jsx

```diff
@@ -186,13 +189,46 @@ Section.defaultProps = {
 export const SectionIntl = injectIntl(Section);
 
 export class _Sections extends React.PureComponent {
+  getOrderedSectionsData() {
+    const sectionsData = [];
+    const enabledSections = this.props.Sections.filter(section => section.enabled);
+    const prefs = this.props.Prefs.values;
+    for (const sectionId of prefs.sectionOrder.split(",")) {
+      if (sectionId === "topsites" && prefs.showTopSites) {
+        sectionsData.push({id: sectionId});
+      } else {
+        const section = enabledSections.find(s => s.id === sectionId);
+        if (section) {
+          sectionsData.push(section);
+        }
+      }
+    }
+    return sectionsData;
+  }
+
+  renderSections() {
+    const sectionsData = this.getOrderedSectionsData();
+    const sections = [];
+    for (let i = 0; i < sectionsData.length; i++) {
+      const section = sectionsData[i];
+      const extraProps = {
+        key: section.id,
+        isFirst: i === 0,
+        isLast: i === sectionsData.length - 1
+      };
+      if (section.id === "topsites") {
+        sections.push(<TopSites {...extraProps} />);
+      } else {
+        sections.push(<SectionIntl {...section} Prefs={this.props.Prefs} dispatch={this.props.dispatch} {...extraProps} />);
+      }
+    }
```

Comment:
while you're looking, if you have any ideas on how to simplify these two functions above 😄 

The issue is that I need to figure out all the sections I am going to render and their order first, to then apply the `isFirst` and `isLast` props. Initially, I was throwing the `<components/>` into the array in one loop but then there was no way to go update the props on them. So I have to loop twice. If that makes sense?

Reply:
You should be able to do something like this from the first/only loop?
```js
{
  isFirst: sectionsData.length === 0,
  isLast: sectionsData.length == enabledSections.length - 1
}
```

And instead of pushing temporary objects, just push the jsx?

Reply:
ooh. hmm. yeah. I guess that works as long as the user doesn't manually mess up the sectionOrder pref. I'm fine with that.